numba/core/types/containers.py

```diff
@@ -393,14 +393,31 @@ def key(self):
 
 
 class NamedTuple(_HeterogeneousTuple, BaseNamedTuple):
-    def __init__(self, types, cls):
+    def __new__(cls, types, clss, convert=True):
+        """
+        arg convert: whether allow NamedTuple convert to Uni, default True.
+        """
+        if not convert:
+            return object.__new__(NamedTuple)
+
+        t = utils.unified_function_type(types, require_precise=True)
```

Comment:
If I'm not mistaken this line would allow the casting of one type into another, prioritizing the creation of NamedUniTuple over maintaning the original types. I might be wrong but I think that if the input is (int64, float64), then the types will be unified as (float64, float64) and then a NamedUniTuple will be created. Shouldn't we prefer to preserve the original types by using a NamedTuple in this case? And only create NamedUniTuple when the types are the same.

Reply:
@dlee992 following on from the discussion at the Numba meeting, I'd guess that the types are being constructed differently for some reason, this is done here for maps: https://github.com/numba/numba/blob/4684b9cf9d5e42ec0e144c3f238432b078627a1c/numba/core/typeinfer.py#L317-L326 and here for lists: https://github.com/numba/numba/blob/4684b9cf9d5e42ec0e144c3f238432b078627a1c/numba/core/typeinfer.py#L278-L285 it's the `__call__` methods that do the construction logic. That might be the place to start debugging.

Reply:
@stuartarchibald , thanks! I will debug from here ASAP.

Reply:
I was wrong in my comment above. That line will only apply to function types, and it will not affect other types like float or int.

Reply:
@luk-f-a , thanks for ur new finding. @stuartarchibald, I tried with ur suggestions. However, these two test cases don't call into the `call` method of `BuildListConstraint`. I now tried to debug about `InsertValue` and `TupleModel` stuff, since they are in the code location when error happens.

Reply:
I think I have found the root cause. Since I change `NamedTuple` could be converted to `NamedUniTuple` by `new` method, but `LiteralStrKeyDict` is registered to use `TupleModel`. Actually, it should can also use `UniTupleModel`, since sometimes `LiteralStrKeyDict` will use `NamedUniTuple` as underlying datastruct.